clients/unshaded/src/main/java/tachyon/client/block/BufferedBlockInStream.java

```diff
@@ -0,0 +1,218 @@
+  protected long mBufferPos;
+  /** Current position of the stream, relative to the start of the block. */
+  protected long mPos;
+
+  /**
+   * Helper for creating a BufferedBlockInStream. This sets the necessary variables and creates
+   * the initial buffer which is empty and invalid.
+   *
+   * @param blockId block id for this stream
+   * @param blockSize size of the block in bytes
+   * @param location worker address to read the block from
+   */
+  // TODO: Get the block lock here when the remote instream locks at a stream level
+  public BufferedBlockInStream(long blockId, long blockSize, InetSocketAddress location) {
+    mBlockId = blockId;
+    mBlockSize = blockSize;
+    mLocation = location;
+    mBuffer = allocateBuffer();
+    mBufferPos = INVALID_BUFFER_POS; // No data in buffer
+    mClosed = false;
+    mContext = BlockStoreContext.INSTANCE;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public int read() throws IOException {
+    checkIfClosed();
+    if (remaining() == 0) {
+      close();
+      return -1;
+    }
+    if (remainingInBuffer() == 0) {
+      updateBuffer();
+    }
+    mPos ++;
+    mBufferPos ++;
+    return BufferUtils.byteToInt(mBuffer.get());
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(b != null, "Read buffer cannot be null");
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        String.format("Buffer length (%d), offset(%d), len(%d)", b.length, off, len));
+    if (len == 0) {
+      return 0;
+    } else if (remaining() == 0) { // End of block
+      return -1;
+    }
+
+    int toRead = (int) Math.min(len, remaining());
+    if (remainingInBuffer() > toRead) { // data is fully contained in the buffer
+      mBuffer.get(b, off, toRead);
+      mPos += toRead;
+      mBufferPos += toRead;
+      return toRead;
+    }
+
+    if (toRead > mBuffer.limit() / 2) { // directly read if request is > one-half buffer size
+      mBufferPos = INVALID_BUFFER_POS; // Invalidate the buffer
+      int bytesRead = directRead(b, off, toRead);
+      mPos += bytesRead;
+      return bytesRead;
+    }
+
+    // For a read <= half the buffer size, fill the buffer first, then read from the buffer.
+    updateBuffer();
+    mBuffer.get(b, off, toRead);
+    mPos += toRead;
+    mBufferPos += toRead;
+    return toRead;
+  }
+
+  @Override
+  public long remaining() {
+    return mBlockSize - mPos;
+  }
+
+  @Override
+  public void seek(long pos) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(pos >= 0, "Seek position is negative: " + pos);
+    Preconditions.checkArgument(pos <= mBlockSize, "Seek position is past end of block: "
```

Comment:
I think you can seek to `blockSize`, it will just be moving the marker to the end of the stream, so any read will return -1.